src/pages/reader/epub.vue

```diff
@@ -233,155 +406,183 @@ export default {
+      if (!this.searchText.length) return;
+
+      const searchResultsByChapters = await Promise.all(
+        this.book.spine.spineItems.map(this.searchChapter)
       );
-      return results.flat().slice(0, 1000);
-    },
-    updateSearchResults(searchResults = []) {
-      this.searchResults = searchResults;
-      this.searchResults.forEach(result => {
-        this.rendition.annotations.highlight(result.cfi);
-      });
-    },
-    saveToLocalStorage(currentCfi) {
-      if (window.localStorage && currentCfi) {
-        window.localStorage.setItem(
-          `epub-reader-${this.cacheKey}`,
-          JSON.stringify({ currentCfi: currentCfi.toString() })
-        );
-      }
-    },
-    resumeFromLocalStorage() {
-      if (window.localStorage) {
-        const epubData = window.localStorage.getItem(
-          `epub-reader-${this.cacheKey}`
-        );
-        if (epubData) {
-          try {
-            const { currentCfi } = JSON.parse(epubData);
-            return currentCfi;
-          } catch (error) {
-            // eslint-disable-next-line no-console
-            console.error(error);
-          }
+      this.searchResults = searchResultsByChapters.flat();
+      if (!this.searchResults.length) return;
+
+      this.selectedSearchResultIndex = this.findSelectedSearchResultIndex();
+
+      this.focusSelectedSearchResult();
+    }, 1000),
+    findSelectedSearchResultIndex() {
+      if (!this.searchResults.length) return 0;
+
+      // Find the first search result that is after the current location
+      let start = 0;
+      let end = this.searchResults.length - 1;
+      while (start <= end) {
+        const mid = Math.floor((start + end) / 2);
+        const cfi = this.searchResults[mid];
+        const { compareStart } = this.compareCFIWithCurrent(cfi);
+
+        if (compareStart === 0) {
+          return mid;
+        }
+        if (compareStart < 0) {
+          start = mid + 1;
+        } else {
+          end = mid - 1;
         }
       }
-      return undefined;
+
+      return start < this.searchResults.length ? start : 0;
     },
-    async directToSelectedSearchResult() {
+    async focusSelectedSearchResult() {
       if (!this.searchResults.length) return;
+
       const cfiString = this.searchResults[
-        this.searchResultIndex
-      ].cfi.toString();
+        this.selectedSearchResultIndex
+      ].toString();
       await this.rendition.display(cfiString);
       const range = this.rendition.getRange(cfiString);
       const selection = this.contents.window.getSelection();
-      // TODO: selection is null when the page is not rendered (e.g. change chapter)
+      // Selection becomes null when changing chapter
       if (selection) {
         selection.removeAllRanges();
         selection.addRange(range);
+      } else {
+        // Focus again after changing chapter
+        this.$nextTick(() => {
+          this.focusSelectedSearchResult();
```

Comment:
Would this cause infinite loop somehow though

Reply:
Use highlight instead selection for selected search result